cishouseholds/pipeline/generate_outputs.py

```diff
@@ -25,6 +27,67 @@
 from cishouseholds.pipeline.pipeline_stages import register_pipeline_stage
 
 
+@register_pipeline_stage("report")
+def report(
+    unique_id_column: str,
+    error_column: str,
+    duplicate_row_flag_column: str,
+    processed_file_table: str,
+    invalid_files_table: str,
+    valid_survey_table: str,
+    invalid_survey_table: str,
+    output_directory: str,
+):
+    valid_df = extract_from_table(valid_survey_table)
+    invalid_df = extract_from_table(invalid_survey_table)
+    processed_file_log = extract_from_table(processed_file_table)
+    invalid_files_log = extract_from_table(invalid_files_table)
+    processed_file_count = processed_file_log.count()
+    invalid_files_count = invalid_files_log.count()
+    num_valid_survey_responses = valid_df.count()
+    num_invalid_survey_responses = invalid_df.count()
+    valid_df_errors = valid_df.select(unique_id_column, error_column)
+    invalid_df_errors = invalid_df.select(unique_id_column, error_column)
+
+    valid_df_errors = valid_df_errors.withColumn("ERROR LIST", F.explode(error_column)).groupBy("ERROR LIST").count()
+    invalid_df_errors = (
+        invalid_df_errors.withColumn("ERROR LIST", F.explode(error_column)).groupBy("ERROR LIST").count()
+    )
+
+    duplicated_df = valid_df.filter(F.col(duplicate_row_flag_column) > 1).union(
+        valid_df.filter(F.col(duplicate_row_flag_column) > 1)
+    )
+
+    counts_df = pd.DataFrame(
+        {
+            "dataset": ["processed_file_log", "invalid_file_log", "valid_survey_responses", "invalid_survey_responses"],
+            "count": [
+                processed_file_count,
+                invalid_files_count,
+                num_valid_survey_responses,
+                num_invalid_survey_responses,
+            ],
+        }
+    )
+    with pd.ExcelWriter("report_output.xlsx") as writer:
+        counts_df.to_excel(writer, sheet_name="dataset totals", index=False)
+        valid_df_errors.toPandas().to_excel(writer, sheet_name="errors in valid dataset", index=False)
+        invalid_df_errors.toPandas().to_excel(writer, sheet_name="errors in invalid dataset", index=False)
+        duplicated_df.toPandas().to_excel(writer, sheet_name="duplicated rows", index=False)
```

Comment:
Rather than writing the file locally to copy across, could you [write it to a bytes object in memory](https://stackoverflow.com/a/56976741/8103477) and then pass that to `write_string_to_file`?